angelsrefining/prototypes/angels-functions.lua

```diff
@@ -930,6 +1076,202 @@ function angelsmods.functions.check_exception(arg, excep)
   end
   return true
 end
+-------------------------------------------------------------------------------
+-- FLUID FLOW AND RECIPE COLOUR FUNCTIONS -------------------------------------
+-------------------------------------------------------------------------------
+--the following functions were gratefully donated by @Maxreader on Discord
+local function isColor(input)
+  if type(input) ~= "table" then
+    return nil
+    end
+    local length = table_size(input)
+    if length ~= 3 and length ~= 4 then
+        return nil
+    end
+    local is_small = true
+    for _,v in pairs(input) do 
+        if v > 1 then
+            is_small = false
+            break
+        end
+    end
+    if not is_small then
+        for k,v in pairs(input) do
+            input[k] = v/255
+        end
+    end
+  local color = {
+    r = input.r or input[1],
+    g = input.g or input[2],
+    b = input.b or input[3],
+    a = input.a or input[4] or 1
+  }
+  return color
+end

```

Comment:
@maxreader indented my feelings
```suggestion
local function isColor(input)
  if type(input) ~= "table" then
    return nil
  end
  local length = table_size(input)
  if length ~= 3 and length ~= 4 then
    return nil
  end
  local is_small = true
  for _,v in pairs(input) do 
    if v > 1 then
      is_small = false
      break
    end
  end
  if not is_small then
    for k,v in pairs(input) do
      input[k] = v/255
    end
  end
  local color = {
    r = input.r or input[1],
    g = input.g or input[2],
    b = input.b or input[3],
    a = input.a or input[4] or 1
  }
  return color
end
```